NBitcoin/Protocol/Node.cs

```diff
@@ -1074,41 +1087,57 @@ public void RespondToHandShake(CancellationToken cancellation = default(Cancella
 			}
 		}
 
-		public void Disconnect()
-		{
-			Disconnect(null, null);
-		}
-
 		int _Disconnecting;
 
-		public void Disconnect(string reason, Exception exception = null)
+		/// <summary>
+		/// Disconnects the node and checks the listener thread.
+		/// </summary>
+		public void Disconnect(string reason = null)
 		{
-			DisconnectAsync(reason, exception);
-			AssertNoListeningThread();
-			_Connection.Disconnected.WaitOne();
-		}
+			if (IsConnected == false)
+				return;
 
-		private void AssertNoListeningThread()
-		{
-			if(_Connection._ListenerThreadId == Thread.CurrentThread.ManagedThreadId)
-				throw new InvalidOperationException("Using Disconnect on this thread would result in a deadlock, use DisconnectAsync instead");
+			DisconnectNode(reason, null);
+
+			try
+			{
+				AssertListenerThreadIsNotCurrentThread();
+				_Connection.Disconnected.WaitOne();
+			}
+			catch (InvalidOperationException)
+			{
+				throw;
+			}
+			finally
+			{
+				_Connection.CleanUp();
+			}
 		}
-		public void DisconnectAsync()
+
+		/// <summary>
+		/// Disconnects the node without checking the listener thread.
+		/// </summary>
+		public void DisconnectAsync(string reason = null, Exception exception = null)
 		{
-			DisconnectAsync(null, null);
+			if(IsConnected == false)
+				return;
+
+			DisconnectNode(reason, exception);
+			_Connection.CleanUp();
 		}
-		public void DisconnectAsync(string reason, Exception exception = null)
+
+		private void DisconnectNode(string reason = null, Exception exception = null)
 		{
-			if(!IsConnected)

```

Comment:
No need for this?

Reply:
Shouldn't we check if the node is connected before we disconnect?

Reply:
We should but seems you have remove it no?